packages/xsnap/src/xsnap.js

```diff
@@ -340,35 +411,112 @@ export async function xsnap(options) {
   }
 
   /**
-   * @param {string} [description]
+   * @param {string} description
+   * @param {import('@endo/promise-kit').PromiseKit<void>} batonKit
    * @returns {AsyncGenerator<Uint8Array, void, undefined>}
    */
-  async function* makeSnapshot(description = 'unknown') {
-    // TODO: Refactor to use tmpFile rather than tmpName.
-    const tmpSnapPath = await ptmpName({
-      template: `make-snapshot-${safeHintFromDescription(
-        description,
-      )}-XXXXXX.xss`,
-    });
-
+  async function* makeSnapshotInternal(description, batonKit) {
+    const output = new PassThrough({ highWaterMark: 1024 * 1024 });
+    let piped = false;
+    let cleaned = false;
+    let done = Promise.resolve();
+
+    let snapshotReadSize = 0;
+    /** @type {number | undefined} */
+    let snapshotSize;
     try {
-      const result = baton.then(async () => {
-        await messagesToXsnap.next(encoder.encode(`w${tmpSnapPath}`));
-        await runToIdle();
+      /** @type {string} */
+      let snapPath;
+      /** @type {Readable} */
+      let sourceStream;
+
+      const maybePipe = () => {
+        if (!piped && !cleaned) {
+          sourceStream.pipe(output);
+          piped = true;
+        }
+      };
+
+      if (snapshotUseFs) {
+        // TODO: Refactor to use tmpFile rather than tmpName.
+        // eslint-disable-next-line @jessie.js/no-nested-await
+        snapPath = await ptmpName({
+          template: `make-snapshot-${safeHintFromDescription(
+            description,
+          )}-XXXXXX.xss`,
+        });
+        // eslint-disable-next-line @jessie.js/no-nested-await
+        const handle = await fs.open(snapPath, 'w+');
```

Comment:
Why `w+` ("writing and reading"). It sounds like this is the filename we send to the `xsnap` "please write a snapshot" command, which means `xsnap` will write to it, and we'll read from it. So should it use `r` instead of `w+`?

(using multiple modes on a single file is suspicious)

Reply:
This is a bit of stream opening race mitigation. Before I'd create the file name, tell xsnap write here (which it does with `wb`), wait for the response, then create the stream, wait for the file open event, then hook the stream.

To simplify the logic, I'm abusing file modes, and now the flow is as follow: JS creates the file in `w+` mode, aka create/truncate and allow reading, creates the read stream, tell xsnap to take the snapshot, wait for the response, and then connects the stream to the output. 

It's a lot more similar to the wiring for the streaming snapshot, except for when the piping of source to output stream occurs. If not that, I'd have to create a thunk to call in the "maybePipe", with all the asynchronous error handling complications.

I can try to document this better.

Reply:
Ah, ok, so it's a bit like what `tail -F` must do (wait for the file to exist, *then* wait for new data to appear). But here, the kernel side creates the file in the first place (so it can start waiting for new data right away), and the worker side opens a pre-existing file for writing (i.e. truncate, but the reader doesn't have to wait for them).

And of course Node's enthusiasm for async means that it's one notch more complicated: the kernel side can't just open the file, it has to start opening a file and then wait for the open to complete.

I'd be tempted to do a sync `open(mode='w')` to create the empty file, then close it right away, then an async `open(mode='r')` to set up the reader, with a comment in between to explain that we pre-create the file to avoid needing to watch for it to appear when the child starts writing. But I can imagine how that flow might be not as good as the `mode='w+'` you're doing.

So yeah, please add a comment explaining what the two sides observe, and why `w+` is a tolerable abuse.